ValueCoder fails with NegativeArraySizeException if serialized size exceeds 2GB

# Affected Daml version

1.18.1

# To reproduce

Run this program (as a scala test test case):
```
    val n = 500000000
    val textValue = ValueText("a" * n)
    val recordValue = ValueRecord(
      None,
      ImmArray(
        None -> textValue,
        None -> textValue,
        None -> textValue,
        None -> textValue,
        None -> textValue,
      ),
    )
    val versionedValue = Versioned(TransactionVersion.VDev, recordValue)

    // Fails with NegativeArraySizeException
    val valuePOrErr = ValueCoder.encodeVersionedValue(ValueCoder.CidEncoder, versionedValue)

    // Expected behavior: The program should fail with an EncodeError with a human readable message
    val valueP = valuePOrErr.value

    Console.println(valueP.getValue.size())
```

Alternative scenario: Set n to 900000000. Then it fails with an EncodeError, but the message is a bit clunky:
```
Serializing com.daml.lf.value.ValueOuterClass$Value to a ByteString threw an IOException (should never happen).
java.lang.RuntimeException: Serializing com.daml.lf.value.ValueOuterClass$Value to a ByteString threw an IOException (should never happen).
	at com.google.protobuf.AbstractMessageLite.toByteString(AbstractMessageLite.java:62)
	at com.daml.lf.value.ValueCoder$.encodeValue(ValueCoder.scala:517)
	at com.daml.lf.value.ValueCoder$.encodeVersionedValue(ValueCoder.scala:388)
	at com.daml.lf.value.ValueCoder$.encodeVersionedValue(ValueCoder.scala:380)
	at com.digitalasset.canton.protocol.WellFormedTransactionTest.$anonfun$new$9(WellFormedTransactionTest.scala:388)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.wordspec.AnyWordSpecLike$$anon$3.apply(AnyWordSpecLike.scala:1227)
	at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
	at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
	at org.scalatest.wordspec.AnyWordSpec.withFixture(AnyWordSpec.scala:1879)
	at org.scalatest.wordspec.AnyWordSpecLike.invokeWithFixture$1(AnyWordSpecLike.scala:1225)
	at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$runTest$1(AnyWordSpecLike.scala:1237)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
	at org.scalatest.wordspec.AnyWordSpecLike.runTest(AnyWordSpecLike.scala:1237)
	at org.scalatest.wordspec.AnyWordSpecLike.runTest$(AnyWordSpecLike.scala:1219)
	at org.scalatest.wordspec.AnyWordSpec.runTest(AnyWordSpec.scala:1879)
	at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$runTests$1(AnyWordSpecLike.scala:1296)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
	at org.scalatest.wordspec.AnyWordSpecLike.runTests(AnyWordSpecLike.scala:1296)
	at org.scalatest.wordspec.AnyWordSpecLike.runTests$(AnyWordSpecLike.scala:1295)
	at org.scalatest.wordspec.AnyWordSpec.runTests(AnyWordSpec.scala:1879)
	at org.scalatest.Suite.run(Suite.scala:1112)
	at org.scalatest.Suite.run$(Suite.scala:1094)
	at org.scalatest.wordspec.AnyWordSpec.org$scalatest$wordspec$AnyWordSpecLike$$super$run(AnyWordSpec.scala:1879)
	at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$run$1(AnyWordSpecLike.scala:1341)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
	at org.scalatest.wordspec.AnyWordSpecLike.run(AnyWordSpecLike.scala:1341)
	at org.scalatest.wordspec.AnyWordSpecLike.run$(AnyWordSpecLike.scala:1339)
	at com.digitalasset.canton.protocol.WellFormedTransactionTest.org$scalatest$BeforeAndAfterAll$$super$run(WellFormedTransactionTest.scala:17)
	at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
	at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
	at com.digitalasset.canton.protocol.WellFormedTransactionTest.run(WellFormedTransactionTest.scala:17)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:45)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13(Runner.scala:1322)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13$adapted(Runner.scala:1316)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1316)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24(Runner.scala:993)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24$adapted(Runner.scala:971)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1482)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:971)
	at org.scalatest.tools.Runner$.run(Runner.scala:798)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2or3(ScalaTestRunner.java:38)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:25)
Caused by: com.google.protobuf.CodedOutputStream$OutOfSpaceException: CodedOutputStream was writing to a flat byte array and ran out of space.
	at com.google.protobuf.CodedOutputStream$ArrayEncoder.writeStringNoTag(CodedOutputStream.java:1506)
	at com.google.protobuf.CodedOutputStream$ArrayEncoder.writeString(CodedOutputStream.java:1216)
	at com.google.protobuf.GeneratedMessageV3.writeString(GeneratedMessageV3.java:3090)
	at com.daml.lf.value.ValueOuterClass$Value.writeTo(ValueOuterClass.java:1904)
	at com.google.protobuf.CodedOutputStream$ArrayEncoder.writeMessageNoTag(CodedOutputStream.java:1306)
	at com.google.protobuf.CodedOutputStream$ArrayEncoder.writeMessage(CodedOutputStream.java:1274)
	at com.daml.lf.value.ValueOuterClass$RecordField.writeTo(ValueOuterClass.java:5811)
	at com.google.protobuf.CodedOutputStream$ArrayEncoder.writeMessageNoTag(CodedOutputStream.java:1306)
	at com.google.protobuf.CodedOutputStream$ArrayEncoder.writeMessage(CodedOutputStream.java:1274)
	at com.daml.lf.value.ValueOuterClass$Record.writeTo(ValueOuterClass.java:4784)
	at com.google.protobuf.CodedOutputStream$ArrayEncoder.writeMessageNoTag(CodedOutputStream.java:1306)
	at com.google.protobuf.CodedOutputStream$ArrayEncoder.writeMessage(CodedOutputStream.java:1274)
	at com.daml.lf.value.ValueOuterClass$Value.writeTo(ValueOuterClass.java:1888)
	at com.google.protobuf.AbstractMessageLite.toByteString(AbstractMessageLite.java:59)
	... 54 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: Failed writing a at index 205032799
	at com.google.protobuf.Utf8$UnsafeProcessor.encodeUtf8(Utf8.java:1526)
	at com.google.protobuf.Utf8.encode(Utf8.java:293)
	at com.google.protobuf.CodedOutputStream$ArrayEncoder.writeStringNoTag(CodedOutputStream.java:1487)
	... 67 more
```

# Additional context

Protobuf is known to not support messages beyond 2GB. Unfortunately, Protobuf seems to fail ungracefully on the attempt to serialize an excessively large message.

It would be very helpful, if you could explore how Protobuf fails if the message is too big and map these exceptions to human readable EncodeErrors.

--- a/daml-lf/transaction/BUILD.bazel
+++ b/daml-lf/transaction/BUILD.bazel
@@ -58,7 +58,33 @@ da_scala_library(
         "//daml-lf/data",
         "//daml-lf/language",
         "//libs-scala/nameof",
+        "//libs-scala/safe-proto",
         "//libs-scala/scala-utils",
+        "@maven//:com_google_protobuf_protobuf_java",
+    ],
+)
+
+da_scala_test(
+    name = "value-test",
+    srcs = glob([
+        "src/test/**/value/*.scala",
+        "src/test/**/EitherAssertions.scala",
+    ]),
+    max_heap_size = "3g",
+    scala_deps = [
+        "@maven//:com_chuusai_shapeless",
+        "@maven//:org_scalacheck_scalacheck",
+        "@maven//:org_scalatestplus_scalacheck_1_15",
+        "@maven//:org_scalaz_scalaz_core",
+        "@maven//:org_scalaz_scalaz_scalacheck_binding",
+    ],
+    scalacopts = lf_scalacopts,
+    deps = [
+        ":transaction",
+        ":value_proto_java",
+        "//daml-lf/data",
+        "//daml-lf/interface",
+        "//daml-lf/transaction-test-lib",
         "@maven//:com_google_protobuf_protobuf_java",
     ],
 )
@@ -69,7 +95,6 @@ da_scala_test(
     srcs = glob([
         "src/test/**/EitherAssertions.scala",
         "src/test/**/crypto/*.scala",
-        "src/test/**/value/*.scala",
         "src/test/**/transaction/*.scala",
     ]),
     scala_deps = [
@@ -77,7 +102,6 @@ da_scala_test(
         "@maven//:org_scalacheck_scalacheck",
         "@maven//:org_scalatestplus_scalacheck_1_15",
         "@maven//:org_scalaz_scalaz_core",
-        "@maven//:org_scalaz_scalaz_scalacheck_binding",
     ],
     scalacopts = lf_scalacopts,
     deps = [

--- a/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala
+++ b/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala
@@ -1,7 +1,8 @@
 // Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
 // SPDX-License-Identifier: Apache-2.0
 
-package com.daml.lf
+package com.daml
+package lf
 package value
 
 import com.daml.lf.data.Ref._
@@ -385,10 +386,10 @@ object ValueCoder {
       value: Value,
   ): Either[EncodeError, proto.VersionedValue] =
     for {
-      protoValue <- encodeValue(encodeCid, version, value)
+      bytes <- encodeValue(encodeCid, version, value)
     } yield {
       val builder = proto.VersionedValue.newBuilder()
-      builder.setVersion(encodeValueVersion(version)).setValue(protoValue).build()
+      builder.setVersion(encodeValueVersion(version)).setValue(bytes).build()
     }
 
   /** Serialize a Value to protobuf
@@ -514,7 +515,7 @@ object ValueCoder {
     }
 
     try {
-      Right(go(0, v0).toByteString)
+      SafeProto.toByteString(go(0, v0)).left.map(EncodeError)
     } catch {
       case Err(msg) => Left(EncodeError(msg))
     }
@@ -526,4 +527,5 @@ object ValueCoder {
   ): Either[DecodeError, Value] = {
     decodeValue(decodeCid, proto.VersionedValue.parseFrom(bytes))
   }
+
 }

--- a/daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/value/ValueCoderSpec.scala
+++ b/daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/value/ValueCoderSpec.scala
@@ -1,7 +1,8 @@
 // Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
 // SPDX-License-Identifier: Apache-2.0
 
-package com.daml.lf
+package com.daml
+package lf
 package value
 
 import com.daml.lf.EitherAssertions
@@ -27,6 +28,19 @@ class ValueCoderSpec
 
   implicit override val generatorDrivenConfig: PropertyCheckConfiguration =
     PropertyCheckConfiguration(minSuccessful = 1000)
+
+  "encode" should {
+    "fail gracefully when serialized message exceeding 2GB" in {
+
+      val ver = TransactionVersion.StableVersions.max
+
+      val value0 = ValueText("a" * (1024 * 1024 * 1024))
+      val value1 = ValueList(FrontStack(value0, value0))
+
+      ValueCoder.encodeValue(ValueCoder.CidEncoder, ver, value0) shouldBe a[Right[_, _]]
+      ValueCoder.encodeValue(ValueCoder.CidEncoder, ver, value1) shouldBe a[Left[_, _]]
+    }
+  }
 
   "encode-decode" should {
     "do Int" in {

--- a/libs-scala/safe-proto/BUILD.bazel
+++ b/libs-scala/safe-proto/BUILD.bazel
@@ -1,0 +1,26 @@
+# Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+load(
+    "//bazel_tools:scala.bzl",
+    "da_scala_library",
+    "da_scala_test_suite",
+)
+
+da_scala_library(
+    name = "safe-proto",
+    srcs = glob(["src/main/scala/**/*.scala"]),
+    tags = ["maven_coordinates=com.daml:safe-proto:__VERSION__"],
+    visibility = ["//visibility:public"],
+    deps = ["@maven//:com_google_protobuf_protobuf_java"],
+)
+
+da_scala_test_suite(
+    name = "safe-protot-test",
+    srcs = glob(["src/test/scala/**/*.scala"]),
+    max_heap_size = "4g",
+    deps = [
+        ":safe-proto",
+        "@maven//:com_google_protobuf_protobuf_java",
+    ],
+)

--- a/libs-scala/safe-proto/src/main/scala/com/daml/SafeProto.scala
+++ b/libs-scala/safe-proto/src/main/scala/com/daml/SafeProto.scala
@@ -1,0 +1,39 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml
+
+import com.google.protobuf.{AbstractMessageLite, ByteString, CodedOutputStream}
+
+object SafeProto {
+
+  // `AbstractMessageLite#toByteString` and `AbstractMessageLite#toByteArray` on
+  // messages requiring more than 2GB fails with two different RuntimeExceptions
+  // which are not human readable. The root cause of the failure is that
+  // AbstractMessageLite#getSerializedSize computation does not handle overflow.
+  // We identified two different error cases when `getSerializedSize` overflows:
+  //  - if `getSerializedSize` overflows and outputs a positive int,
+  //    the serialization throws a generic `RuntimeException` with a
+  //    `CodedOutputStream.OutOfSpaceException` as cause.
+  //  - if `getSerializedSize` overflows and outputs a negative int,
+  //    the serialization throws a `NegativeArraySizeException`
+  // This function catches those exceptions.
+  private[this] def safelySerialize[X](
+      message: AbstractMessageLite[_, _],
+      serialize: AbstractMessageLite[_, _] => X,
+  ): Either[String, X] =
+    try {
+      Right(serialize(message))
+    } catch {
+      case e: RuntimeException
+          if e.isInstanceOf[NegativeArraySizeException] ||
+            e.getCause != null && e.getCause.isInstanceOf[CodedOutputStream.OutOfSpaceException] =>
+        Left(s"the ${message.getClass.getName} message is too big to be serialized")
+    }
+
+  def toByteString(message: AbstractMessageLite[_, _]): Either[String, ByteString] =
+    safelySerialize(message, _.toByteString)
+
+  def toByteArray(message: AbstractMessageLite[_, _]): Either[String, Array[Byte]] =
+    safelySerialize(message, _.toByteArray)
+}

--- a/libs-scala/safe-proto/src/test/scala/com/daml/SafeProtoTest.scala
+++ b/libs-scala/safe-proto/src/test/scala/com/daml/SafeProtoTest.scala
@@ -1,0 +1,41 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml
+
+import com.google.protobuf._
+import org.scalatest.Inside
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+class SafeProtoTest extends AnyWordSpec with Inside with Matchers {
+
+  private[this] def tuple(v: Value) =
+    Value.newBuilder().setListValue(ListValue.newBuilder().addValues(v).addValues(v)).build()
+
+  private[this] val value0 = Value.newBuilder().setStringValue("a" * 1024 * 1024 * 1024).build()
+  assert(value0.getSerializedSize > 0)
+  private[this] val value1 = tuple(value0)
+  assert(value1.getSerializedSize < 0)
+  private[this] val value2 = tuple(value1)
+  assert(value2.getSerializedSize >= 0)
+
+  "toByteString" should {
+
+    "fail gracefully if the message to serialize exceeds 2GB" in {
+      SafeProto.toByteString(value0) shouldBe a[Right[_, _]]
+      SafeProto.toByteString(value1) shouldBe a[Left[_, _]]
+      SafeProto.toByteString(value2) shouldBe a[Left[_, _]]
+    }
+  }
+
+  "toByteArray" should {
+
+    "fail gracefully if the message to serialize exceeds 2GB" in {
+      SafeProto.toByteArray(value0) shouldBe a[Right[_, _]]
+      SafeProto.toByteArray(value1) shouldBe a[Left[_, _]]
+      SafeProto.toByteArray(value2) shouldBe a[Left[_, _]]
+    }
+  }
+
+}

--- a/release/artifacts.yaml
+++ b/release/artifacts.yaml
@@ -207,6 +207,8 @@
   type: jar-scala
 - target: //libs-scala/resources-grpc:resources-grpc
   type: jar-scala
+- target: //libs-scala/safe-proto:safe-proto
+  type: jar-scala
 - target: //libs-scala/scala-utils:scala-utils
   type: jar-scala
 - target: //libs-scala/scalatest-utils:scalatest-utils

